### ISSUE TITLE
Clarify that template is published/unpublished

### ISSUE THREAD

**Issue by stephaniesimms** [DMPRoadmap/roadmap#251]
When I click Edit template details and check the box to 'Publish' the template, then click Save, the template is published but the message still reads 'false' when it should be 'true.' I only know that the template was successfully published when I return to the main templates page and see that published status is 'Yes' (screenshots below).

![screen shot 2017-04-14 at 11 59 34 am](https://cloud.githubusercontent.com/assets/12703638/25053044/defbae22-2109-11e7-9e6b-d9c5e6d92670.png)

![screen shot 2017-04-14 at 11 53 06 am](https://cloud.githubusercontent.com/assets/12703638/25053034/c7776dea-2109-11e7-8f6c-c4366b32478f.png)

![screen shot 2017-04-14 at 11 53 40 am](https://cloud.githubusercontent.com/assets/12703638/25053051/e343eeea-2109-11e7-9a2a-973f530eec7a.png)

**Comment by briri** [comment]
I think this may be a larger problem with the way the system is handling versioning. 

@stephaniesimms @sjDCC have either of you do any testing of template versioning? 
For example, update a template and then verifying whether or not existing plans continue to see the old version and if new plans get the latest version? Or if you edit a published template that a new plan doesn't see your new changes until you have republished?

**Comment by sjDCC** [comment]
There are definitely errors on template publishing.

I've just created a new test template and tried to publish this. The publishing checkbox under 'template details' seems to save (I get a changed saved message bar) but the template shows as unpublished under template details and as published "Yes (unsaved changes)" under the main table in the templates tab. 

I tried to create a plan on the front end to see whether it was actually published and got a 404 message when trying to pick up this template (i.e. by picking no funder)

Unclear to me whether it's published or not currently.

![capture](https://cloud.githubusercontent.com/assets/6169940/25234679/c660c520-25da-11e7-82ce-edb04bd7b765.PNG)
![capture2](https://cloud.githubusercontent.com/assets/6169940/25234697/ced909ec-25da-11e7-8170-3ff6f079a49f.PNG)

**Comment by briri** [comment]
There were more significant issues with the logic in the templates_controller. It was creating a separate version every time the template was updated. It was not versioning when changes were made to the phases,sections or questions (regardless of whether or not it was published) and the logic used to determine if the template had "Unpublished changes" was using updated_at and was prone to errors.

I simplified the template logic by:
- Adding a `dirty` flag to the templates table. This flag gets set when the template details change or when a phase, section or question is added/updated/deleted
- Added 3 helper methods to the template model: 
  - `dmptemplate_ids(Org)` returns all of the template version family ids (dmptemplate_id) for the given org
  - `current(Org, dmptemplate_id)` returns the current/latest version of the template 
  - `live(Org, dmptemplate_id)` returns the current published version of the template or nil if the template is not published
- Added 3 new controller endpoints to simplify the logic currently in the `admin_template` method:  
  - `admin_publish` publishes the specified version IF its the current version. It then creates a new version which become the current version to which future changes will be made. The published/live version becomes available to users during plan creation
  - `admin_unpublish` un-publishes the live version of the template. The template is no longer available to users during plan creation and will remain so until the template is republished
  - `admin_customize` creates a customization of a funder template for the current user's org. The customized copy's phases/sections/questions are marked as unmodifiable so that the user cannot change the funder's info
- The pages now check the new `dirty` flag to determine if their are unpublished changes 
- The main list of templates page now only shows the current versions of the org's templates and any funder templates (showing the customized version if it exists)
- New versions are only ever created when a template is published.
- Updated the functional, unit, and integration tests to work with the new logic

**Comment by briri** [comment]
@vyruss, @xsrust  and @jollopre can one or two of you review the changes outlined in this PR https://github.com/CDLUC3/roadmap/pull/46? I made these template changes (outlined above) in the CDLUC3 fork and the easiest way to isolate all of them for you to review was by creating a PR within the CDLUC3 fork. No need to accept and merge the PR, just review the changes for me.
Thanks

**Comment by stephaniesimms** [comment]
I no longer see any option to publish/unpublish a template in the create/edit template pages. There is a "Publish" button in the Actions column on the main templates page, which tells me that my template is published (screenshot below). However, there is no "Stephanie's test template" when I try to create a new plan.

Overall, template data is highly irregular and appears to have multiple issues related to the migration from last Friday...

![screen shot 2017-05-01 at 3 54 31 pm](https://cloud.githubusercontent.com/assets/12703638/25597911/890fa740-2e86-11e7-9366-47e612588fda.png)

**Comment by briri** [comment]
This stems largely from issues with the latest data migration. Every version of every template is published and orgs can have multiple org types due to the use of a bit flag for that field, so an org can be a funder and an institution which makes it hard to determine programmatically whether or not a template is a funder template or an org template

**Comment by briri** [comment]
made some minor modifications to the controller to address some of these issues. will need to wait for better data though to do more thorough testing

**Comment by sjDCC** [comment]
@briri and I discussed where to put the publish button and thought it made most sense in the overview table rather than within the phase. I think that's the only place it exists now.

One question that may help with the issue with the visibility of published templates. What was your institutional affiliation when adding / publishing the template @stephaniesimms? Was it a real university ('institution') or a generic organisation like DCC or DataOne that doesn't have end users based within it? 

Organisations don't show up in the second dropdown under create plan, so that may be why you couldn't get a published template. We've been encouraging organisations to create guidance not templates. Perhaps we should remove the template functionality though if it shows?

**Comment by sjDCC** [comment]
I think there's still something wrong with publishing @briri This is partly in the history table so is perhaps just a display issue, but maybe also in terms of validation (i.e. what shows on the front end) and keeping an accurate record of what was published when?

I created a new template and published it. On doing so, the history table shows two versions (0 and 1) but is noting the earlier one as published. Is that right? It seems counter-intuitive
![data-error](https://cloud.githubusercontent.com/assets/6169940/26007473/85aca9d0-3738-11e7-816a-b6dff4351f61.PNG)

I then unpublished this template. The table updated correctly, but it still let me create a plan with it on the front end. It was recorded everywhere as unpublished so shouldn't have been accessible.
![table2](https://cloud.githubusercontent.com/assets/6169940/26007520/c030073c-3738-11e7-8104-9d02de47f905.PNG)

I added another section to the template and re-published. I wasn't alerted to unsaved changes but I guess I shouldn't be as the template was unpublished at the time. After saving, the history table shows the middle version as the one that is published. Again maybe this is just a display error and the 'published' tag is assigning to the wrong row?
![3rd-test](https://cloud.githubusercontent.com/assets/6169940/26007636/590d8ae2-3739-11e7-89f4-3f339522ba23.PNG)

When you follow the links in this table to view the different versions, the last two are identical. It seems that the content of a template is not frozen once published. You make changes, which get flagged as unpublished changes, but the act of saving these replaces the old published version and creates an identical new version. Does that make sense? This isn't ideal as you lose the sense of what was published at any given time

**Comment by briri** [comment]
When you publish a template it locks/publishes the current version and then creates a new 'working' copy that's why it always looks like the one prior has been published. We could switch it up so that 

Apologies for not updating the plan creation section to prevent unpublish templates from being listed. I figured it best to wait though since I was starting on the new plan creation page next. The new page will only show published templates, so that should be a non-issue once that page is in place (we should definitely test though).

In your example above when you published for the 2nd time it published/locked version 1 and created a version 2. Its an identical copy, so I can see the confusion there. When you make changes to version 2 though it definitely should NOT be updating the published version though.

I think we can take 2 approaches here. Update the history page so that its less confusing, or change the timing of the new version creation. If someone goes in to edit a published template we can create the new version at that time.

**Comment by sjDCC** [comment]
Thanks @briri. Glad my comments made sense. 

I think we probably want to change the table so it doesn't list the new working copy. Or maybe add a note / help icon to clarify it's a working draft? @JEK-III or @stephaniesimms may have suggestions.

No need to update plan creation section now. If unpublished stuff will filter out later that's fine.

**Comment by vyruss** [comment]
@briri this change creates the following problem: because it's in the _template_ update action, this effectively lets you edit a published template if you _save_ the phase or a question. Therefore in order to prevent this, the template's other components' update actions (phase update, section update etc) need to be changed too, to create a new template version every time they're saved and this entails quite a change - so here's what I've done:

1. User clicks on Edit -> this creates a (`deep_copy`) working copy and takes them to it for editing.
2. This makes the published version no longer editable _(which is semantically correct because you can never edit a template already in use)_
3. The history table now shows the working copy suffixed with _(pencil-icon) Draft_

@sjDCC does that look ok?
before edit: ![before_edit](https://cloud.githubusercontent.com/assets/18707439/26161536/1aa7fc4c-3b1c-11e7-8502-8e6a12970368.png)
after edit: ![after_edit](https://cloud.githubusercontent.com/assets/18707439/26161542/1e3bbe3e-3b1c-11e7-8f9a-d5584e1289ed.png)

**Comment by briri** [comment]
I think this solution could work @vyruss. At what point are you creating the new version? On the `admin_template` method instead of the `admin_update` in the controller?

**Comment by vyruss** [comment]
@briri exactly. I'll push shortly

**Comment by sjDCC** [comment]
Sorry. I'm only just getting chance to look at this now. 

Definitely good to lock templates for editing once published. Our user message tells people they're viewing a historical version so it makes sense to grey it all out / make read only. 

I assume the template draft gets created once someone goes into edit a published template i.e. by clicking in one of these two places
![capture](https://cloud.githubusercontent.com/assets/6169940/26214580/d0f9e6a8-3bf4-11e7-8e71-00ceda3a9e6e.PNG)
![capture2](https://cloud.githubusercontent.com/assets/6169940/26214579/d0f78322-3bf4-11e7-9e9b-f3f8ded4cffb.PNG)

That makes sense to me and I particularly like the pencil icon :)

**Comment by sjDCC** [comment]
All seems good. Closing out
